### PR TITLE
[DCOS-58675] Implement base test framework

### PR DESCRIPTION
**Pull request by rpalaznik**
Resolves [[DCOS-58675] Implement base test framework](https://jira.mesosphere.com/browse/DCOS-58675)
- Utility methods to install/uninstall spark operator with helm (to be replaced with kudo later) with custom operator image and namespace, submit test jobs, wait for pods to reach certain status
- A simple test to run linear regression spark job

## How to test
Run the following commands
1. Create a cluster with `make cluster-create`
2. Login to docker hub `docker login`
3. Build and push images with `make docker-push DOCKER_REPO_NAME=<your docker hub handle>`
4. Run the tests `make test DOCKER_REPO_NAME=<your docker hub handle>`

### CODE REVIEW

**Comment by rpalaznik** [comment]
> First pass of code looks good! But when I tried to run `make test` on a default `konvoy up` cluster (what we use in our CI), I saw that the linear regression test got stuck due to insufficient cpu resources. What kind of k8s cluster did you use?

Thanks for noticing that! I had used a cluster with 5 worker nodes, instead of 4. I did some changes to make the test work on a default cluster.

File: tests/utils/utils_test.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+package utils
```
Comment (akirillov) [inline tests/utils/utils_test.go]: 👍 for writing tests for test utils!

File: tests/utils/templates.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	file, err := ioutil.TempFile("/tmp", "spark-test-")
+	if err != nil {
+		log.Fatal(err)
+		panic(err)
+	}
+
+	err = templates.ExecuteTemplate(file, sparkRbacTemplateName, map[string]string{"Namespace": namespace})
+	if err != nil {
+		log.Fatal(err)
+		panic(err)
```
Comment (akirillov) [inline tests/utils/templates.go]: not familiar with this library, but can you cross-check if `log.Fatal(err)` calls `panic(err)` after printing the message? So `panic(err)` could be redundant here.

File: tests/run.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+set -e
+
+# The following environment variables are available to customize test execution:
+# - KUBECONFIG
```
Comment (samvantran) [inline tests/run.sh]: hm, for some reason now `make test` for me doesn't work with error 
```
docker run -i --rm \
        -v {#redacted}/kudo-spark-operator/tests:/tests \
        -v {#redacted}/kudo-spark-operator/admin.conf:/root/.kube/config \
        -e KUBECONFIG=/root/.kube/config \
        -e SPARK_IMAGE="mesosphere/spark-2.4.4-bin-hadoop2.7-k8s:2cd2dc39bde0dbcf9c2e76a94e004db01f2779fe" \
        -e SPARK_OPERATOR_IMAGE="mesosphere/kudo-spark-operator:327b0d0bbc4444f01c565c02618dd90c5f2d4a03" \
        mesosphere/spark-operator-docker-builder:b622eeb7b4f15d0297fac1400e8009e25e7c9946 \
        /bin/bash -c \
        "kubectl cluster-info && \
        echo \$SPARK_IMAGE && echo \$SPARK_OPERATOR_IMAGE"
# tests entrypoint
error: error loading config file "/root/.kube/config": read /root/.kube/config: is a directory
make: *** [Makefile:89: test] Error 1```

Looks like it's missing a mount of the kube config somewhere?

**Comment by akirillov** [comment]
Thanks, @rpalaznik. I retried the tests and the tooling worked out of the box! `TestSparkOperatorInstallation` passed but `TestSparkOperatorInstallationWithCustomNamespace` got stuck:
```
kubectl get pods
NAME                                                 READY   STATUS      RESTARTS   AGE
kudo-spark-operator-sparkoperator-74cdbd9465-zxf8x   1/1     Running     0          4m34s
kudo-spark-operator-sparkoperator-init-7xlxl         0/1     Completed   0          4m34s
linear-regression-1570164825151-exec-1               1/1     Running     0          4m13s
linear-regression-1570164825151-exec-2               0/1     Pending     0          4m13s
linear-regression-driver                             1/1     Running     0          4m22s
```
So probably it makes sense to change executor instances from `2` to `1` in `spark-linear-regression-job.yaml`?

After abnormal termination of the test runner, consequent `make test` failed:
```
Running tests
mesosphere/kudo-spark-operator/testsmesosphere/kudo-spark-operator
=== RUN   TestSparkOperatorInstallation
INFO[2019-10-03T22:03:39-07:00] Installing Spark Operator with helm
INFO[2019-10-03T22:03:39-07:00] Configuring RBAC
INFO[2019-10-03T22:03:39-07:00] Applying file /tmp/spark-test-665460900 with kubectl
INFO[2019-10-03T22:03:40-07:00] kubectl output:
INFO[2019-10-03T22:03:40-07:00] namespace/kudo-spark-operator-testing unchanged
serviceaccount/spark-driver unchanged
role.rbac.authorization.k8s.io/spark-driver-role unchanged
rolebinding.rbac.authorization.k8s.io/spark-driver-role-binding unchanged
INFO[2019-10-03T22:03:40-07:00] Adding the repository
INFO[2019-10-03T22:03:40-07:00] Installing the chart
FATA[2019-10-03T22:03:43-07:00] exit status 1
FAIL	github.com/mesosphere/kudo-spark-operator/tests	4.043s
FAIL
make: *** [Makefile:67: test] Error 1
```
Shall we introduce cleanup step prior to running tests?

**Comment by rpalaznik** [comment]
> can we somehow make logging more verbose when calling shell commands

It should be better now.

> So probably it makes sense to change executor instances from `2` to `1` in `spark-linear-regression-job.yaml`?

I did decrease the number of executors to 1 in this commit https://github.com/mesosphere/kudo-spark-operator/pull/6/commits/f6f921d4f3c6f7e1a204d3fc0b50dd0303dea94a

> After abnormal termination of the test runner, consequent `make test` failed:
> ...
> Shall we introduce a cleanup step prior to running tests?

It definitely makes sense to do it, but maybe it should be implemented after a switch to kudo operator installation in the tests?

File: tests/run.sh (hunk 3 above)
Comment (rpalaznik) [reply tests/run.sh]: Have you tried it with `make cluster-create`? 
If not, when using a pre-existing cluster, try setting `KUBECONFIG` explicitly when calling `make test`

**Pull request merged**